src/audio/igo_nr/igo_nr.c

```diff
@@ -0,0 +1,521 @@
+	}
+
+	for (i = 0; i < sink_frames; i++) {
+		y = audio_stream_write_frag_s32(sink, idx_out);
+		*y = (cd->out[cd->out_rpt]);
+
+		if (cd->config.igo_params.dump_data == 1 && out_nch > 1) {
+			y = audio_stream_write_frag_s32(sink, idx_out + 1);
+			*y = (cd->in[cd->out_rpt]);
+		}
+
+		cd->out_rpt++;
+		idx_out += out_nch;
+	}
+}
+
+static void igo_nr_capture_s32(struct igo_nr_comp_data *cd,
+			       const struct audio_stream *source,
+			       struct audio_stream *sink,
+			       int32_t src_frames,
+			       int32_t sink_frames)
+{
+	int32_t in_nch = source->channels;
+	int32_t out_nch = sink->channels;
+	int32_t i;
+	int32_t idx_in = 0;
+	int32_t idx_out = 0;
+	int32_t *x;
+	int32_t *y;
+
+	for (i = 0; i < src_frames; i++) {
+		x = audio_stream_read_frag_s32(source, idx_in);
+		cd->in[cd->in_wpt] = (*x) >> 16;
+
+		cd->in_wpt++;
+		idx_in += in_nch;
+	}
+
+	for (i = 0; i < sink_frames; i++) {
+		y = audio_stream_write_frag_s32(sink, idx_out);
+		*y = (cd->out[cd->out_rpt]) << 16;
+
+		if (cd->config.igo_params.dump_data == 1 && out_nch > 1) {
+			y = audio_stream_write_frag_s32(sink, idx_out + 1);
+			*y = (cd->in[cd->out_rpt]) << 16;
+		}
+
+		cd->out_rpt++;
+		idx_out += out_nch;
+	}
+}
+
+static inline int32_t set_capture_func(struct comp_dev *dev)
+{
+	struct igo_nr_comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sourceb;
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+
+	switch (sourceb->stream.frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		comp_info(dev, "set_capture_func(), SOF_IPC_FRAME_S16_LE");
+		cd->igo_nr_func = igo_nr_capture_s16;
+		break;
+	case SOF_IPC_FRAME_S32_LE:
+		comp_info(dev, "set_capture_func(), SOF_IPC_FRAME_S32_LE");
+		cd->igo_nr_func = igo_nr_capture_s32;
+		break;
+	default:
+		comp_err(dev, "set_capture_func(), invalid frame_fmt");
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static struct comp_dev *igo_nr_new(const struct comp_driver *drv,
+				   struct sof_ipc_comp *comp)
+{
+	struct sof_ipc_comp_process *ipc_igo_nr =
+		(struct sof_ipc_comp_process *)comp;
+	struct comp_dev *dev;
+	struct igo_nr_comp_data *cd;
+	size_t bs = ipc_igo_nr->size;
+	int32_t ret;
+
+	comp_cl_info(&comp_igo_nr, "igo_nr_new()");
+
+	/* Check first that configuration blob size is sane */
+	if (bs > SOF_IGO_NR_MAX_SIZE) {
+		comp_cl_err(&comp_igo_nr, "igo_nr_new() error: configuration blob size = %u > %d",
+			    bs, SOF_IGO_NR_MAX_SIZE);
+		return NULL;
+	}
+
+	dev = comp_alloc(drv, COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+
+	memcpy_s(COMP_GET_IPC(dev, sof_ipc_comp_process),
+		 sizeof(struct sof_ipc_comp_process), ipc_igo_nr,
+		 sizeof(struct sof_ipc_comp_process));
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	ret = IgoLibGetInfo(&cd->igo_lib_info);
+	if (ret != IGO_RET_OK) {
+		comp_cl_err(&comp_igo_nr, "igo_nr_new(): IgoLibGetInfo() Failed.");
+		rfree(dev);
```

Comment:
inittialize cd to NULL and you can use the same goto here and above since free(NULL) -> no-op

Reply:
Done. Thanks for the hint!